pkg/controller/volume/persistentvolume/scheduler_binder.go

```diff
@@ -0,0 +1,414 @@
+	boundClaims = []*v1.PersistentVolumeClaim{}
+	unboundClaimsImmediate = []*v1.PersistentVolumeClaim{}
+	unboundClaims = []*bindingInfo{}
+
+	for _, vol := range pod.Spec.Volumes {
+		volumeBound, pvc, err := b.isVolumeBound(pod.Namespace, &vol, false)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		if pvc == nil {
+			continue
+		}
+		if volumeBound {
+			boundClaims = append(boundClaims, pvc)
+		} else {
+			delayBinding, err := b.ctrl.shouldDelayBinding(pvc)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			if delayBinding {
+				// Scheduler path
+				unboundClaims = append(unboundClaims, &bindingInfo{pvc: pvc})
+			} else {
+				// Immediate binding should have already been bound
+				unboundClaimsImmediate = append(unboundClaimsImmediate, pvc)
+			}
+		}
+	}
+	return boundClaims, unboundClaims, unboundClaimsImmediate, nil
+}
+
+func (b *volumeBinder) checkBoundClaims(claims []*v1.PersistentVolumeClaim, node *v1.Node, podName string) (bool, error) {
+	for _, pvc := range claims {
+		pvName := pvc.Spec.VolumeName
+		pv, err := b.pvCache.GetPV(pvName)
+		if err != nil {
+			return false, err
+		}
+
+		err = volumeutil.CheckNodeAffinity(pv, node.Labels)
+		if err != nil {
+			glog.V(4).Infof("PersistentVolume %q, Node %q mismatch for Pod %q: %v", pvName, node.Name, err.Error(), podName)
+			return false, nil
+		}
+		glog.V(5).Infof("PersistentVolume %q, Node %q matches for Pod %q", pvName, node.Name, podName)
+	}
+
+	glog.V(4).Infof("All volumes for Pod %q match with Node %q", podName, node.Name)
+	return true, nil
+}
+
+func (b *volumeBinder) findMatchingVolumes(podName string, claimsToBind []*bindingInfo, node *v1.Node) (foundMatches bool, err error) {
+	// Sort all the claims by increasing size request to get the smallest fits
+	sort.Sort(byPVCSize(claimsToBind))
+
+	allPVs := b.pvCache.ListPVs()
+	chosenPVs := map[string]*v1.PersistentVolume{}
+
+	for _, bindingInfo := range claimsToBind {
+		// Find a matching PV
+		bindingInfo.pv, err = findMatchingVolume(bindingInfo.pvc, allPVs, node, chosenPVs, true)
+		if err != nil {
+			return false, err
+		}
+		if bindingInfo.pv == nil {
+			glog.V(4).Infof("No matching volumes for PVC %q on node %q", getPVCName(bindingInfo.pvc), node.Name)
+			return false, nil
+		}
+
+		// matching PV needs to be excluded so we don't select it again
+		chosenPVs[bindingInfo.pv.Name] = bindingInfo.pv
+	}
+
+	// Mark cache with all the matches for each PVC for this node
+	b.podBindingCache.UpdateBindings(podName, node.Name, claimsToBind)
+	glog.V(4).Infof("Found matching volumes on node %q", node.Name)
+
+	return true, nil
+}
+
+func (b *volumeBinder) revertAssumedPVs(bindings []*bindingInfo) {
+	for _, bindingInfo := range bindings {
+		b.pvCache.Restore(bindingInfo.pv.Name)
+	}
+}
+
+type bindingInfo struct {
+	// Claim that needs to be bound
+	pvc *v1.PersistentVolumeClaim
+
+	// Proposed PV to bind to this claim
+	pv *v1.PersistentVolume
+}
+
+func (b bindingInfo) String() string {
```

Comment:
Unit test coverage indicates that this is never called. Is it used anywhere?

Reply:
I had added it for debugging, but don't need it anymore. Will remove.

Reply:
I remember now. It's used to print out unit test errors.  Will add comment.